### ISSUE TITLE
seems to not support the case where seq_q is longer than seq_k and (Is_causal || Is_local) == True

### ISSUE THREAD

**Issue by NiuMa-1234** [Dao-AILab/flash-attention#1778]
Hi, Thank you for your maverlous work! I test the case of MHA where seq_q is 2048 and k is 1024 , with Is_causal be True, and found half part of the result is 0. And I further found the n_block_max would be negative under this case, which might be the reason and a bug(?

Here's where the n_block_max is defined in code. When `(seqlen_k - seqlen_q)` is negative, the n_block_max would be negative either, and the second loop for dense attention wouldn't happen (and finally caused the '`seqlen_q -seqlen_k`' rows of the result to remain their initial value: 0 ).
```
if constexpr (Is_causal) {
            n_block_max = std::min(
                n_block_max,
                cute::ceil_div((m_block + 1) * kBlockM_div_H + seqlen_k - seqlen_q, kBlockN));
        }
```

And here is my test script:

```

import math
import pytest
import torch
import torch.nn.functional as F
from einops import rearrange, repeat
from flash_attn import (
    flash_attn_func,
)
from flash_attn.bert_padding import pad_input, unpad_input
from flash_attn.flash_attn_interface import _get_block_size_n
from flash_attn.layers.rotary import apply_rotary_emb

MAX_HEADDIM_SM8x = 192


is_sm75 = torch.cuda.get_device_capability("cuda") == (7, 5)
is_sm8x = torch.cuda.get_device_capability("cuda")[0] == 8
is_sm80 = torch.cuda.get_device_capability("cuda") == (8, 0)
is_sm90 = torch.cuda.get_device_capability("cuda") == (9, 0)


def attention_ref(q, k, v, query_padding_mask=None, key_padding_mask=None, dropout_p=0.0,
                  dropout_mask=None, causal=False, bias=None, upcast=True, reorder_ops=False, diy_blockmasks=None):
    """
    Arguments:
        q: (batch_size, seqlen_q, nheads, head_dim)
        k: (batch_size, seqlen_k, nheads, head_dim)
        v: (batch_size, seqlen_k, nheads, head_dim)
    Output:
        output: (batch_size, seqlen_q, nheads, head_dim)
        attention: (batch_size, nheads, seqlen_q, seqlen_k), softmax after dropout
    """
    dtype_og = q.dtype
    if upcast:
        q, k, v = q.float(), k.float(), v.float()
    seqlen_q, seqlen_k = q.shape[1], k.shape[1]
    d = q.shape[-1]
    if not reorder_ops:
        scores = torch.einsum('bthd,bshd->bhts', q / math.sqrt(d), k)
    else:
        scores = torch.einsum('bthd,bshd->bhts', q, k / math.sqrt(d))
    if bias is not None:
        scores = (scores + bias).to(dtype=scores.dtype)
    if key_padding_mask is not None:
        scores.masked_fill_(rearrange(~key_padding_mask, 'b s -> b 1 1 s'), float('-inf'))
    print("scores: ",scores)
    if causal:
        causal_mask = torch.triu(torch.ones(seqlen_q, seqlen_k, dtype=torch.bool, device=q.device), 1)
    else:
        # causal_mask = torch.zeros(seqlen_q, seqlen_k, dtype=torch.bool, device=q.device)
        causal_mask = torch.full((seqlen_q, seqlen_k), False, dtype=torch.bool, device=q.device)
        print("causal_mask: ",causal_mask.shape)
    scores.masked_fill_(causal_mask, float('-inf'))
    print("scores after: ",scores)
    attention = torch.softmax(scores, dim=-1)
    dropout_scaling = 1.0 / (1 - dropout_p)
    # attention_drop = attention.masked_fill(~dropout_mask, 0.0) * dropout_scaling
    # output = torch.einsum('bhts,bshd->bthd', attention_drop , v)
    if dropout_mask is not None:
        attention_drop = attention.masked_fill(~dropout_mask, 0.0)
    else:
        attention_drop = attention
    output = torch.einsum('bhts,bshd->bthd', attention_drop, v * dropout_scaling)
    if query_padding_mask is not None:
        output.masked_fill_(rearrange(~query_padding_mask, 'b s -> b s 1 1'), 0.0)
        attention = attention.masked_fill(rearrange(~query_padding_mask, 'b s -> b 1 s 1'), 0.0)
    return output.to(dtype=dtype_og), attention.to(dtype=dtype_og)


def test_flash_attn_deterministic_orig(seqlen_q, seqlen_k, swap_sq_sk, d, causal, local, dtype):
    if (
        max(seqlen_q, seqlen_k) >= 2048
        and torch.cuda.get_device_properties("cuda").total_memory <= 16 * 2**30
    ):
        pytest.skip()  # Reference implementation OOM
    if swap_sq_sk:
        seqlen_q, seqlen_k = seqlen_k, seqlen_q
    device = "cuda"
    # set seed
    torch.random.manual_seed(0)
    
    batch_size = 1
    nheads = 2
    q = torch.randn(batch_size, seqlen_q, nheads, d, device=device, dtype=dtype, requires_grad=True)
    k = torch.randn(batch_size, seqlen_k, nheads, d, device=device, dtype=dtype, requires_grad=True)
    # v = torch.randn(batch_size, seqlen_k, nheads, d, device=device, dtype=dtype, requires_grad=True)
    v = torch.randint(high=20000,size=(batch_size,seqlen_k,nheads, d), device=device)
    v = v.to(torch.bfloat16) 
    v.requires_grad=True
    out = flash_attn_func(q, k, v, 0.0, causal=causal)

    q_ref, k_ref, v_ref, = q.detach().clone(), k.detach().clone(), v.detach().clone()
    output_ref, attn_ref = attention_ref(q_ref, k_ref, v_ref, None, None,
                                         0.0, dropout_mask=None, causal=causal)
    
    
    
    q_fake, k_fake, v_fake = q.detach().clone(), k.detach().clone(), v.detach().clone()
    fake_output_ref, _ = attention_ref(q_fake, k_fake, v_fake, None, None,
                                         0.0, dropout_mask=None, causal=not causal)
    
    g = torch.randn_like(out)
    dq0, dk0, dv0 = torch.autograd.grad(out, (q, k, v), g, retain_graph=True)
    for _ in range(50):
        dq, dk, dv = torch.autograd.grad(out, (q, k, v), g, retain_graph=True)
        assert torch.equal(dv, dv0)
        assert torch.equal(dk, dk0)
        assert torch.equal(dq, dq0)
    print("pass test!")

if __name__ =="__main__":
    # test_flash_attn_deterministic(4, 3, swap_sq_sk=False, d=32, causal=False, local = True, dtype = torch.bfloat16)
    test_flash_attn_deterministic_orig(4, 3, swap_sq_sk=False, d=32, causal=True, local = True, dtype = torch.bfloat16)
```
And I'm not sure if this is literally a bug and if the changes to this part of code for cases where seq_q is longer than seq_k could cause any new error. Could you please help me ? Many thanks.

**Comment by tridao** [comment]
Because we define the causal mask to be aligned to the bottom right, if seqlen_q > seqlen_k, the top of the output will be zero. That's intended.

https://github.com/Dao-AILab/flash-attention/blob/7321879fde54f09ed94f7f6ce9377e2f4cf1fac0/flash_attn/flash_attn_interface.py#L1082

**Comment by NiuMa-1234** [comment]
Understood. Many thanks!
